tokio/src/sync/watch.rs

```diff
@@ -412,18 +418,32 @@ mod state {
         }
 
         /// Load the current value of the state.
-        pub(super) fn load(&self) -> StateSnapshot {
-            StateSnapshot(self.0.load(SeqCst))
+        pub(super) fn load(&self, ordering: Ordering) -> StateSnapshot {
+            StateSnapshot(self.0.load(ordering))
+        }
+
+        /// Load the current value of the state.
+        ///
+        /// The receiver side (read-only) uses `Acquire` ordering for a proper handover
+        /// with the sender side (single writer).
+        pub(super) fn load_receiver(&self) -> StateSnapshot {
+            StateSnapshot(self.0.load(Ordering::Acquire))
         }
 
         /// Increment the version counter.
         pub(super) fn increment_version(&self) {
-            self.0.fetch_add(STEP_SIZE, SeqCst);
+            // Use `Release` ordering to ensure that storing the version
+            // state is seen by the receiver side that uses `Acquire` for
+            // loading the state.
+            self.0.fetch_add(STEP_SIZE, Ordering::Release);
         }
 
         /// Set the closed bit in the state.
         pub(super) fn set_closed(&self) {
-            self.0.fetch_or(CLOSED_BIT, SeqCst);
+            // Use `Release` ordering to ensure that storing the version
+            // state is seen by the receiver side that uses `Acquire` for
+            // loading the state.
+            self.0.fetch_or(CLOSED_BIT, Ordering::Release);
```

Comment:
That said, it would arguably make sense to use something stronger than relaxed for `ref_count_rx`. Currently this assert could fail:
```
// thread 1
GLOBAL_BOOL.store(true, Relaxed);
drop(receiver);

// thread 2
if sender.is_closed() {
    assert!(GLOBAL_BOOL.load(Relaxed));
}
```

Reply:
Both the sender and all receivers could be dropped independently in any order by different threads. I can't imagine how to link the sender-side *closed* state (refcount) with the receiver-side *closed* state (version bit) when using separate atomics.

Reply:
Either way, it's a separate question from what you're fixing here.